src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -78,20 +78,13 @@ private void StopRaisingEvents()
                 {
                     // Stop the RunLoop and wait for the thread to exit gracefully
                     Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
-                    _watcherThread.Join();
-                    _watcherRunLoop = IntPtr.Zero;
                 }
 
-                // Clean up the EventStream, if it exists
-                if (!_eventStream.IsInvalid)
-                {
-                    _eventStream = new SafeEventStreamHandle(IntPtr.Zero);
-                }             
-
-                // Cleanup the callback
-                if (_callback != null)
+                // Gurd against us from calling Join on our own background thread (for 
+                // example, calling Dispose or Stop from a callback).
+                if (Thread.CurrentThread != _watcherThread)
                 {
-                    _callback = null;
+                    _watcherThread.Join();
```

Comment:
Can this blocking be moved outside of the _syncLock?

Otherwise it seems like there's still potentially a possible deadlock, e.g. if a callback try to start the watcher again, which will require taking the lock, which can't be acquired until the watcher thread shuts down, which won't happen until the callback returns, and deadlock.

I'm thinking:
```C#
Thread watcherThread;
lock (_syncLock)
{
    watcher = _watcherThread;
    _watcherThread = null;
    ...
}
if (watcher != null && watcher != Thread.CurrentThread)
{
    watcher.Join();
}
```
All that said, why are we waiting for this thread at all?  Doesn't look like we do that on either Windows or Linux.  Can we just get rid of this join altogether?